raft/src/main/java/org/apache/kafka/raft/KafkaRaftClient.java

```diff
@@ -1700,16 +1710,16 @@ private void handleResponse(RaftResponse.Inbound response, long currentTimeMs) {
     }
 
     /**
-     * Validate a request which is only valid between voters. If an error is
-     * present in the returned value, it should be returned in the response.
+     * Validate common state for requests to establish leadership.
+     *
+     * These include the Vote, BeginQuorumEpoch rnd EndQuorumEpoch RPCs. If an error is present in
+     * the returned value, it should be returned in the response.
      */
     private Optional<Errors> validateVoterOnlyRequest(int remoteNodeId, int requestEpoch) {
         if (requestEpoch < quorum.epoch()) {
             return Optional.of(Errors.FENCED_LEADER_EPOCH);
         } else if (remoteNodeId < 0) {
             return Optional.of(Errors.INVALID_REQUEST);
-        } else if (quorum.isObserver() || !quorum.isVoter(remoteNodeId)) {
-            return Optional.of(Errors.INCONSISTENT_VOTER_SET);
```

Comment:
In KIP-853, `INCONSISTEN_VOTER_SET` is deprecated and replicas will not return this error anymore.

In this case replicas that think they are observer need to be allowed to vote if the leader thinks they are voters. This can happen if a voter is added to the set of voters right before an election cycle and the VotersRecord has been replicated to the new voter.